sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/datetimeExpressions.scala

```diff
@@ -2067,6 +2082,10 @@ object DatePart {
        224
       > SELECT _FUNC_('SECONDS', timestamp'2019-10-01 00:00:01.000001');
        1.000001
+      > SELECT _FUNC_('days', interval 1 year 10 months 5 days);
```

Comment:
So, feel free to tell me to go look up a reference, but I am sort of confused about the semantics. The "days" part of "interval 1 year 10 months 5 days" is 5 days, not the interval days (roughly 365 + 10 *31 + 5).

But below, the "seconds" part of "30 seconds 1 ms 1 us" is 30.001001, not 30.

So what's the number of days in "interval 1 year 10 months 5 days 12 hours"? 5 or 5.5?

Reply:
Sort of answering my own question. From PostgreSQL, at least:
https://www.postgresql.org/docs/9.1/functions-datetime.html#FUNCTIONS-DATETIME-EXTRACT

```
SELECT date_part('day', TIMESTAMP '2001-02-16 20:38:40');
Result: 16

SELECT date_part('hour', INTERVAL '4 hours 3 minutes');
Result: 4
```

It seems like the answer to the second example here should be 30?

I'm getting off on a tangent, but, can you specify "interval 5 minutes 90 seconds"? if so, what's the minute part -- 5 or 6? if you can't specify that, can you specify "interval 90 seconds"? if not why not?

Just getting confused about the intended semantics of the date part of an interval!

Reply:
> So, feel free to tell me to go look up a reference, but I am sort of confused about the semantics.

Semantic is defined by internal representation of Catalyst's `CalendarIntervalType`. Purpose of the type is to represent components of an interval not duration of interval (see the discussions in https://github.com/apache/spark/pull/25022). `CalendarIntervalType` stores the `year`, `months`, `weeks`, `days`, `hours`, `minutes`, `seconds`, `milliseconds`, `microseconds` components in 2 independent variables  - `months` and `microseconds`. Due to independence and unlimited (limited by long/int size) of the `microseconds` field, we can store more than 31 days in `microseconds`. 
Because purpose of `CalendarIntervalType` is to represent interval as components but not duration, the same duration in microseconds can have multiple representations in Catalyst `INTERVAL` type. Let's say:
390000000 microseconds = `interval 5 minutes 90 seconds` = `interval 6 minutes 30 seconds` = `interval 1 minute 330 seconds` and so on. 

When an user request components of an interval, we need to select one of its representation. We do that by constructing "normalized" representation in where number of `years` and `days` is unlimited (limited by int or long) and `months` should be in `[0..12)`, `hours` is in `[0, 12)`, `minutes` is `[0, 60)` ... `microseconds` is in `[0, 1000000)`. `seconds` and `milliseconds` are special cases because they have the fractional part as well but they limited too. `seconds` - `[0, 59.999999]`, milliseconds - `[0, 59999.999]`. This behavior is inherited from PostgreSQL. In this way, there is only one "normalized" representation of any interval from which the `date_part()` function extracts components.

Reply:
Yeah I get why we split the months and ms part, and can see what the convenient implementation is w.r.t. the current implementation. Maybe the details of the semantics aren't that important for practical purposes.

But it does seem like this proposal doesn't match PostgreSQL in the `SELECT date_part('hour', INTERVAL '4 hours 3 minutes');` case, at least. But it's impractical to follow this behavior given the internal representation.

`SELECT date_part('hour', INTERVAL '1 month 1 day');` gives 24, right? because the days are separable from the months.

`SELECT date_part('month', INTERVAL '1 year 1 month');` gives 13, not 1, right? because the month part isn't separable.

This feels inconsistent. What if we construed the semantics to always mean 'the given interval in the given units'? that's consistent, but doesn't quite sound like what `date_part` does, as it's no longer a 'part'.

Am I right about that so far?

Reply:
> But it does seem like this proposal doesn't match PostgreSQL in the `SELECT date_part('hour', INTERVAL '4 hours 3 minutes')`;

It does. In PostgreSQL:
```sql
maxim=# SELECT date_part('hour', INTERVAL '4 hours 3 minutes');
 date_part 
-----------
         4
```
In SparkSQL:
```sql
spark-sql> SELECT date_part('hour', interval 4 hours 3 minutes);
4
```

> SELECT date_part('hour', INTERVAL '1 month 1 day'); gives 24, right? because the days are separable from the months.

No, it is 0 because the range of hours is `[0, 23]`:
In PostgreSQL:
```sql
maxim=# SELECT date_part('hour', INTERVAL '1 month 1 day');
 date_part 
-----------
         0
```
In SparkSQL:
```sql
spark-sql> SELECT date_part('hour', INTERVAL 1 month 1 day);
0
```

> SELECT date_part('month', INTERVAL '1 year 1 month'); gives 13, not 1, right? because the month part isn't separable.

No, it gives 1.
In PostgreSQL:
```sql
maxim=# SELECT date_part('month', INTERVAL '1 year 1 month');
 date_part 
-----------
         1
```
In SparkSQL:
```sql
spark-sql> SELECT date_part('month', INTERVAL 1 year 1 month);
1
```

> This feels inconsistent.

I cannot agree with you so far.

Reply:
I see, that's good. So as far as you know it's consistent with PostgreSQL? and correctly returns the time-unit part of an interval, not the interval in the time-unit? in that case I'm not sure why this is correct:

```
date_part('seconds', interval 30 seconds 1 milliseconds 1 microseconds);
       30.001001
```

Shouldn't it be 30?

Reply:
> Shouldn't it be 30?

I agree with you that `seconds` should be `30`, `milliseconds` should be `1`, and `microseconds` is `1` but PostgreSQL developers decided to implement this in different way. Maybe this is because `seconds`, `milliseconds` are always have the fractional part in other PostgreSQL functions.

For `seconds`, it looks reasonable but I don't know what PostgreSQL developers thought of when they defined `milliseconds` as milliseconds in a minute:
```sql
maxim=# select date_part('milliseconds', interval '30 seconds 1 milliseconds 1 microseconds');
 date_part 
-----------
 30001.001
``` 
```sql
maxim=# select date_part('microseconds', interval '30 seconds 1 milliseconds 1 microseconds');
 date_part 
-----------
  30001001
```

I interpret this as the `second`, `millisecond`, `microsecond` form one interval component with values in the range `[0, 6000000)` where 0-1 digits belongs to seconds, 2-4 digits to milliseconds, and 5-7 digits to microseconds. 

Precisely follow PostgreSQL implementation is still open question for me. I believe PostgreSQL has badly designed date-time API. As you can see similar `date_part('HOUR', <interval>)` and `date_part('MILLISECONDS', <interval>)` must have similar semantic but they don't.

Reply:
Yeah, it seems inconsistent. Sometimes the entire interval's length is return in the given unit, sometimes just a portion of the interval matching that unit is returned. I wonder if there is any other standard to look at.

Doesn't this current PR have the same consistency question, given the current test output?

I'm mostly concerned with having consistent semantics, whatever they are. If it can't be done reasonably, hm, should we implement this?

I'm trying to figure out the use case for "a date part of an interval". I can only think of cases where the interval should be converted entirely into some unit.

Reply:
> Sometimes the entire interval's length is return in the given unit ...

The entire interval's length is returned for `EPOCH` only, and components for other `field` values.

> I wonder if there is any other standard to look at.

I like java 8 approach where there are 2 types `java.time.Duration` and `java.time.Period`. The first one is to store interval duration in nanosecond precision, the second one is to store components `year`, `months`, `days`. I would reuse this model and extend the types slightly:
- `DURATION` type is to store interval duration in microseconds. `long` should be enough to store difference between any supported timestamps.
- `PERIOD` type should store `years`, `months`, `days`, `hours`, `minutes`, `seconds`, `milliseconds` and `microseconds`. For example, (-10 years, 5 months, -3 hours, 100 microseconds).

> I'm mostly concerned with having consistent semantics, whatever they are. If it can't be done reasonably, hm, should we implement this?

I don't know. From my point of semantic of each extraction field is well defined. There is a difference in implementation of this PR and PostgreSQL. As I wrote above Spark store interval in 2 variables - `months` and `microseconds`, but PostgreSQL uses 3 vars - `month`, `day` and `time` (in microseonds). In this way, `days` are independent from other components. And as consequence, `hours` are not limited by `[0, 24)`:
```sql
maxim=# SELECT interval '-100 years 100 days -100 hours'; 
            interval             
---------------------------------
 -100 years +100 days -100:00:00
(1 row)
```

> I'm trying to figure out the use case for "a date part of an interval". I can only think of cases where the interval should be converted entirely into some unit.

I can image at least 2 use cases:
- Adjusting some of timestamp/date components. For example, we `timestamp have '2019-10-08 10:11:12.123456'`, and we want to adjust it by `-1 year 10 hours`. As result, we will have `timestamp '2018-10-08 20:11:12.123456'`. The `PERIOD` type could be useful here.
- When we need `Ordered` (and comparable) intervals. We can calculate absolute interval duration in some units. For example, in the query:
```sql
SELECT name, place
FROM t1, t2
WHERE t1.sellTimestamp - t2.deliveredTimestamp < interval 1 month 15 days;
```
Here, the `DURATION` should be used because we cannot use `PERIOD` because its values cannot be ordered. Spark's CalendarIntervalType cannot be used here too.

Reply:
> The entire interval's length is returned for `EPOCH` only, and components for other `field` values.

Hm, but then back to the original question; why is the answer not 30 below?
```
SELECT _FUNC_('seconds', interval 30 seconds 1 milliseconds 1 microseconds);
       30.001001
```

I'm not concerned necessarily about matching PostgreSQL, but am concerned about internal consistency. Unless I'm really missing something, these two examples here seem to have different semantics. I understand if you mean it's the natural way to implement this given the representation, but, I'm not sure that makes it right?

I think the discussion of `PERIOD` is separate. Here the question is the use case for `date_part` on an interval. I'm sort of struggling to think of when that's useful, though I accept it exists in other DBs. Your example doesn't show using `date_part`. I can imagine wanting to render an interval in given units -- what is the length of this whole interval in seconds? -- but that's not what this does, nor what the function name seems to imply.

That's why I'm questioning supporting this at all.